htlcswitch/hop/payload.go

```diff
@@ -124,31 +130,52 @@ func NewPayloadFromReader(r io.Reader) (*Payload, error) {
 
 	parsedTypes, err := tlvStream.DecodeWithParsedTypes(r)
 	if err != nil {
-		// Promote any required type failures into ErrInvalidPayload.
-		if e, required := err.(tlv.ErrUnknownRequiredType); required {
-			// If the parser returned an unknown required type
-			// failure, we'll first check that the payload is
-			// properly formed according to our known set of
-			// constraints. If an error is discovered, this
-			// overrides the required type failure.
-			nextHop := lnwire.NewShortChanIDFromInt(cid)
-			err = ValidateParsedPayloadTypes(parsedTypes, nextHop)
-			if err != nil {
-				return nil, err
-			}
-
-			// Otherwise the known constraints were applied
-			// successfully, report the invalid type failure
-			// returned by the parser.
-			return nil, ErrInvalidPayload{
-				Type:      tlv.Type(e),
-				Violation: RequiredViolation,
-				FinalHop:  nextHop == Exit,
-			}
-		}
 		return nil, err
 	}
 
+	// Check for unrecognized mandatory (even) fields and save the lowest
+	// required type to return a deterministic failure.
+	var (
+		requiredViolation        bool
+		minRequiredViolationType tlv.Type
+	)
+	for t, known := range parsedTypes {
+		// If a type is even but not known to us, we cannot process the
+		// payload. We are required to understand a field that we don't
+		// support.
+		//
+		// We always accept custom fields, because a higher level
+		// application may understand them.
+		if known || t%2 != 0 || t >= customTypeStart {
+			continue
+		}
+
+		if !requiredViolation || t < minRequiredViolationType {
+			minRequiredViolationType = t
+		}
+		requiredViolation = true
+	}
+
+	if requiredViolation {
+		// If the parser encountered an unknown required type failure,
+		// we'll first check that the payload is properly formed
+		// according to our known set of constraints. If an error is
+		// discovered, this overrides the required type failure.
+		nextHop := lnwire.NewShortChanIDFromInt(cid)
+		err = ValidateParsedPayloadTypes(parsedTypes, nextHop)
```

Comment:
Is it possible to remove this duplicate call? currently we call this in two different paths, but I think now we can do the required checks after a single `ValudateParsedPayloadTypes`

Reply:
Yes indeed, fixed